src/preload.py

```diff
@@ -0,0 +1,342 @@
+    idx = 0
+    while idx < 4 and size >= 1024:
+        idx += 1
+        size /= 1024
+    result = "{:.{}f}".format(size, precision).rstrip("0").rstrip(".")
+    return "{} {}B".format(result, suffixes[idx])
+
+
+def get_offset_and_size(image, partition):
+    output = parted("-s", "-m", image, "unit", "B", "p").stdout.decode("utf8")
+    lines = output.strip().split("\n")
+    for line in lines:
+        if line.startswith("{}:".format(partition)):
+            data = line.split(":")
+            offset = int(data[1][:-1])
+            size = int(data[3][:-1])
+            return offset, size
+
+
+def mount_partition(image, partition, extra_options):
+    offset, size = get_offset_and_size(image, partition)
+    mountpoint = mkdtemp()
+    options = "offset={},sizelimit={}".format(offset, size)
+    if extra_options:
+        options += "," + extra_options
+    mount("-o", options, image, mountpoint)
+    return mountpoint
+
+
+@contextmanager
+def mount_context_manager(image, partition, extra_options=""):
+    mountpoint = mount_partition(image, partition, extra_options)
+    yield mountpoint
+    umount(mountpoint)
+    os.rmdir(mountpoint)
+
+
+def losetup_partition(image, partition):
+    offset, size = get_offset_and_size(image, partition)
+    device = losetup("-f").stdout.decode("utf8").strip()
+    losetup("-o", offset, "--sizelimit", size, device, image)
+    return device
+
+
+@contextmanager
+def losetup_context_manager(image, partition):
+    device = losetup_partition(image, partition)
+    yield device
+    losetup("-d", device)
+
+
+def api(endpoint, params=None, headers=None):
+    url = "{}/{}".format(API_HOST, endpoint)
+    params = params or {}
+    headers = headers or {}
+    if API_TOKEN:
+        headers["Authorization"] = "Bearer {}".format(API_TOKEN)
+    elif API_KEY:
+        params["apikey"] = API_KEY
+    response = get(url, params=params, headers=headers)
+    response.raise_for_status()
+    return response.json()
+
+
+def get_registry_token(image_repo):
+    params = {
+        "service": REGISTRY_HOST,
+        "scope": "repository:{}:pull".format(image_repo),
+    }
+    return api("auth/v1/token", params)["token"]
+
+
+def registry(endpoint, registry_token, headers=None, decode_json=True):
+    headers = headers or {}
+    headers["Authorization"] = "Bearer {}".format(registry_token)
+    url = "https://{}/{}".format(REGISTRY_HOST, endpoint)
+    response = get(url, headers=headers)
+    response.raise_for_status()
+    return response.json() if decode_json else response.content
+
+
+def get_app_data():
+    """Fetches application metadata"""
+    endpoint = "v2/application({})?$expand=environment_variable".format(APP_ID)
+    data = api(endpoint)["d"][0]
+    image_repo = "{app_name}/{commit}".format(**data).lower()
+    image_id = "{}/{}".format(REGISTRY_HOST, image_repo).lower()
+    env = {v["name"]: v["value"] for v in data.get("environment_variable", [])}
+    return {
```

Comment:
@shaunmulligan fixed in https://github.com/resin-io/resin-preload/pull/65/commits/165f22ed53ef4a41008c1c37de32aee570a18c85